staging/src/k8s.io/pod-security-admission/admission/admission.go

```diff
@@ -0,0 +1,537 @@
+		response = a.ValidateNamespace(ctx, attrs)
+	case corev1.Resource("pods"):
+		response = a.ValidatePod(ctx, attrs)
+	default:
+		response = a.ValidatePodController(ctx, attrs)
+	}
+
+	// TODO: record metrics.
+
+	return response
+}
+
+func (a *Admission) ValidateNamespace(ctx context.Context, attrs admission.Attributes) admissionv1.AdmissionResponse {
+	namespace := attrs.GetObject().(*corev1.Namespace)
+	var oldNamespace *corev1.Namespace
+	if attrs.GetOperation() == admission.Update {
+		var ok bool
+		oldNamespace, ok = attrs.GetOldObject().(*corev1.Namespace)
+		if !ok {
+			klog.InfoS("failed to assert old namespace type", "type", reflect.TypeOf(attrs.GetOldObject()))
+			return internalErrorResponse("failed to decode old namespace")
+		}
+	}
+
+	policy, err := a.PolicyToEvaluate(namespace.Labels)
+
+	// if update, check if policy labels changed
+	if oldNamespace != nil {
+		oldPolicy, oldErr := a.PolicyToEvaluate(oldNamespace.Labels)
+
+		// Only validate on diff.
+		if err != nil && err.Error() != oldErr.Error() {
+			return invalidResponse(err.Error())
+		}
+		// Skip validation if the policy is unchanged.
+		if policy.Enforce.Version == oldPolicy.Enforce.Version {
+			return allowedResponse()
+		}
+		// We can skip validation if the enforce policy is the same pinned version and relaxed.
+		if !policy.Enforce.Latest() && policy.Enforce.Version == oldPolicy.Enforce.Version &&
+			api.CompareLevels(policy.Enforce.Level, oldPolicy.Enforce.Level) < 1 {
+			return allowedResponse()
+		}
+	} else if err != nil { // Create
+		return invalidResponse(err.Error())
+	}
+
+	// We only need to test pods if the enforce policy is not privileged.
+	if policy.Enforce.Level == api.LevelPrivileged {
+		return allowedResponse()
+	}
+
+	response := allowedResponse()
+	response.Warnings = a.CheckPodsInNamespace(ctx, namespace.Name, policy.Enforce)
+
+	return response
+}
+
+func (a *Admission) ValidatePod(ctx context.Context, attrs admission.Attributes) admissionv1.AdmissionResponse {
+	// short-circuit on subresources other than ephemeralcontainers
+	if attrs.GetSubresource() != "" && attrs.GetSubresource() != "ephemeralcontainers" {
+		return allowedResponse()
+	}
+	// short-circuit on exempt namespaces and users
+	if a.exemptNamespace(attrs.GetNamespace()) || a.exemptUser(attrs.GetUserInfo().GetName()) {
+		return allowedResponse()
+	}
+
+	pod, ok := attrs.GetObject().(*corev1.Pod)
+	if !ok {
+		klog.InfoS("failed to assert pod type", "type", reflect.TypeOf(attrs.GetObject()))
+		return internalErrorResponse("failed to decode pod")
+	}
+	enforce := true
+	if attrs.GetOperation() == admission.Update {
+		oldPod, ok := attrs.GetOldObject().(*corev1.Pod)
+		if !ok {
+			klog.InfoS("failed to assert old pod type", "type", reflect.TypeOf(attrs.GetOldObject()))
+			return internalErrorResponse("failed to decode old pod")
+		}
+		if !isSignificantPodUpdate(pod, oldPod) {
+			// Nothing we care about changed, so always allow the update.
+			enforce = false
```

Comment:
if nothing significant changed, why don't we return right here?  Audit messages on insignificant pod updates seems a little odd to me.

Reply:
I sort of agree, though this matches what was [specified](https://github.com/kubernetes/enhancements/tree/master/keps/sig-auth/2579-psp-replacement#updates):

> Update requests to Pods and PodTemplate resources will reevaluate the full object against audit & warn policies, independent of which fields are being modified.

Tim, hitting warn/audit on controller updates makes sense to me... not sure I see the value on a pod update of fields we don't care about

Reply:
> I sort of agree, though this matches what was [specified](https://github.com/kubernetes/enhancements/tree/master/keps/sig-auth/2579-psp-replacement#updates):
> 
> > Update requests to Pods and PodTemplate resources will reevaluate the full object against audit & warn policies, independent of which fields are being modified.
> 
> Tim, hitting warn/audit on controller updates makes sense to me... not sure I see the value on a pod update of fields we don't care about

I'd like a more complete explanation of the benefit we're getting by auditing and warning on field updates that aren't inspected by this admission plugin.

Reply:
We definitely discussed this during the design phase, but I can't remember what the arguments were for always doing audit & warn. I'm OK with skipping those entirely.

Reply:
changed to `return allowedReponse()` at this point